networking/eth2/src/main/java/tech/pegasys/teku/networking/eth2/peers/RateTracker.java

```diff
@@ -35,27 +42,94 @@ public RateTracker(
 
   // boundary: if a request comes in and remaining capacity is at least 1, then
   // they can have the objects they request otherwise they get none.
-  public synchronized long popObjectRequests(final long objectCount) {
+  public synchronized Optional<RequestApproval> popObjectRequests(final long objectCount) {
     pruneRequests();
-    if ((peerRateLimit - requestsWithinWindow) <= 0) {
-      return 0L;
+    final UInt64 currentTime = timeProvider.getTimeInSeconds();
+    if ((peerRateLimit - objectsWithinWindow) <= 0) {
+      return Optional.empty();
     }
+    objectsWithinWindow += objectCount;
+    resetRequestId(currentTime);
+    final RequestApproval requestApproval =
+        new RequestApproval.RequestApprovalBuilder()
+            .requestId(newRequestId.getAndIncrement())
+            .timeSeconds(currentTime)
+            .objectCount(objectCount)
+            .build();
+    requestCount.put(requestApproval.getRequestKey(), objectCount);
+    return Optional.of(requestApproval);
+  }
 
-    requestsWithinWindow += objectCount;
-    requestCount.compute(
-        timeProvider.getTimeInSeconds(),
-        (key, val) -> val == null ? objectCount : val + objectCount);
-    return objectCount;
+  private void resetRequestId(UInt64 currentTime) {
+    if (requestCount.keySet().stream()
+        .noneMatch(
+            objectRequestsEntryKey ->
+                objectRequestsEntryKey.getTimeSeconds().equals(currentTime))) {
+      this.newRequestId.set(0);
+    }
+  }
+
+  public synchronized void adjustObjectRequests(
+      final RequestApproval requestApproval, final long returnedObjectsCount) {
+    pruneRequests();
+    if (requestCount.containsKey(requestApproval.getRequestKey())) {
+      final long initialObjectsCount = requestCount.get(requestApproval.getRequestKey());
+      requestCount.put(requestApproval.getRequestKey(), returnedObjectsCount);
+      objectsWithinWindow = objectsWithinWindow - initialObjectsCount + returnedObjectsCount;
+    }
   }
 
   void pruneRequests() {
     final UInt64 currentTime = timeProvider.getTimeInSeconds();
     if (currentTime.isLessThan(timeoutSeconds)) {
       return;
     }
-    final NavigableMap<UInt64, Long> headMap =
-        requestCount.headMap(currentTime.minus(timeoutSeconds), false);
-    headMap.values().forEach(value -> requestsWithinWindow -= value);
+    final NavigableMap<ObjectRequestsEntryKey, Long> headMap =
+        requestCount.headMap(
+            new ObjectRequestsEntryKey(currentTime.minus(timeoutSeconds), 0), false);
+    headMap.values().forEach(value -> objectsWithinWindow -= value);
     headMap.clear();
   }
+
+  public static class ObjectRequestsEntryKey implements Comparable<ObjectRequestsEntryKey> {
```

Comment:
Maybe RequestKey? Or something other more elegant

Reply:
Done. Renamed it to `RequestKey`